keel-ec2-plugin/src/main/kotlin/com/netflix/spinnaker/keel/ec2/jackson/ClusterSpecDeserializer.kt

```diff
@@ -0,0 +1,44 @@
+package com.netflix.spinnaker.keel.ec2.jackson
+
+import com.fasterxml.jackson.core.TreeNode
+import com.fasterxml.jackson.databind.DeserializationContext
+import com.fasterxml.jackson.databind.JsonNode
+import com.fasterxml.jackson.databind.deser.std.StdNodeBasedDeserializer
+import com.fasterxml.jackson.databind.node.ObjectNode
+import com.netflix.spinnaker.keel.api.RedBlack
+import com.netflix.spinnaker.keel.api.ec2.ClusterSpec
+import com.netflix.spinnaker.keel.api.ec2.ClusterSpec.ServerGroupSpec
+
+class ClusterSpecDeserializer : StdNodeBasedDeserializer<ClusterSpec>(ClusterSpec::class.java) {
+  override fun convert(root: JsonNode, context: DeserializationContext): ClusterSpec =
+    with(context) {
+      ClusterSpec(
+        moniker = treeToValue(root.get("moniker")) ?: error("moniker is required"),
+        imageProvider = treeToValue(root.get("imageProvider")),
+        deployWith = treeToValue(root.get("deployWith")) ?: RedBlack(),
+        locations = treeToValue(root.get("locations")) ?: error("locations is required"),
+        _defaults = ServerGroupSpec(
+          launchConfiguration = treeToValue(root.get("launchConfiguration")),
+          capacity = treeToValue(root.get("capacity")),
+          dependencies = treeToValue(root.get("dependencies")),
+          health = treeToValue(root.get("health")),
+          scaling = treeToValue(root.get("scaling")),
+          tags = treeToValue(root.get("tags"))
+        ),
+        // this is pretty hairy but we can't just use treeToValue because the map's value type is erased
+        overrides = (root.get("overrides") as? ObjectNode?)
+          ?.fields()
+          ?.asSequence()
+          ?.associate { it.key to treeToValue<ServerGroupSpec>(it.value) }
+          ?.filterNotNullValues()
+          ?: emptyMap()
+      )
+    }
```

Comment:
This is doing the work the `@JsonCreator` on `ClusterSpec` used to do -- allowing the default `ServerGroupSpec` properties to be inlined at the top level of the `ClusterSpec`. Longer term I'm wondering if we want to refactor the way that is constructed so that maybe `ClusterSpec` and `ServerGroupSpec` implement a common interface or something. It would certainly remove the need for all this kind of thing, as well as the complexity of `ClusterSpecSchemaCustomizer`.